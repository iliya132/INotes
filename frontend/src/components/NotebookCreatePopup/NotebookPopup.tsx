import React, { useState } from 'react';
import Button from '../Button';
import { IFormInputProps } from './types';
import styles from './NotebookCreatePopup.scss';
import { Input } from '../Input/Input';
import notesController from '../../controllers/NotesController';

export function NotebookPopup(props: IFormInputProps) {
    const { afterClick } = props;
    const [value, setValue] = useState('');

    const handleChange = (event: React.ChangeEvent<HTMLInputElement>) => {
        setValue(event.target.value);
    };

    const handleSubmit = () => {
        notesController.createNotebook(value);
        afterClick();
    };

    return (
        <div className={styles['notebook-create-popup-container']}>
            <Input
                type={'text'}
                id="notebook-name"
                placeholder="Введите имя для новой записной книжки"
                onChange={handleChange}
                data-testid="test-id"
            />
            <Button title="Создать" className={styles['notebook-create-btn']} onClick={handleSubmit} data-testid="test-btn-id"/>
        </div>
    );
}
