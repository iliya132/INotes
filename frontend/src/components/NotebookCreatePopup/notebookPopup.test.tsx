/* eslint-disable no-undef */
import React from 'react';
import { createRoot, Root } from 'react-dom/client';
import { act } from 'react-dom/test-utils';
import userEvent from '@testing-library/user-event';
import { NotebookPopup } from './NotebookPopup';

//#region common setup
let root: Root | null = null;
let container: Element | null = null;
let answers: string[] = [];
jest.mock('../../controllers/NotesController', () => {
    return{
        createNotebook: (title: string) => {answers.push(title)}
    }
});
beforeEach(() => {
    act(() => {
        container = document.createElement('div');
        document.body.appendChild(container);
        root = createRoot(container);
    });
});

afterEach(() => {
    act(() => {
        root!.unmount();
        container!.remove();
        container = null;
    });
});
//#endregion

describe('notebook popup tests', () => {
    it('can be rendered', () => {
        act(() => {
            root!.render(<NotebookPopup afterClick={() => {}} />);
        });
    });

    it('matches snapshot', () => {
        act(() => {
            root!.render(<NotebookPopup afterClick={() => {}} />);
        });

        const popupContainer = container!.querySelector('.notebook-create-popup-container');
        expect(popupContainer).toMatchSnapshot('notebook-popup-snapshot');
    });

    it('contains input', () => {
        act(() => {
            root!.render(<NotebookPopup afterClick={() => {}} />);
        });

        const input = container!.querySelector('input');
        expect(input).not.toBeNull();
        expect(input!.placeholder).toEqual('Введите имя для новой записной книжки');
        expect(input!.id).toEqual('notebook-name');
        expect(input!.type).toEqual('text');
    });

    it('call callback on button pressed', () => {
        let isClicked = false;
        act(() => {
            root!.render(
                <NotebookPopup
                    afterClick={() => {
                        isClicked = true;
                    }}
                />,
            );
        });

        expect(isClicked).toEqual(false);
        const btnElement = container!.querySelector('button');
        expect(btnElement).not.toBeNull();

        act(() => {
            btnElement?.dispatchEvent(new MouseEvent('click', { bubbles: true }));
        });

        expect(isClicked).toEqual(true);
    });

    it('calls notesController.createNotebook', async () => {
        const user = userEvent.setup();

        act(() => {
            root!.render(<NotebookPopup afterClick={() => {}} />);
        });

        const input = container?.querySelector('input')!;
        const btn = container!.querySelector('button')!;

        await act(async () => {
            await user.click(input);
            await user.keyboard("test-input");
            await user.click(btn);
        });

        expect(input.value).toEqual("test-input");
        expect(answers).toContain('test-input');
    });
});
