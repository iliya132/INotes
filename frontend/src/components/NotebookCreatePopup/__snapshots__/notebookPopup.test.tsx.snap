// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`notebook popup tests matches snapshot: notebook-popup-snapshot 1`] = `
<div
  class="notebook-create-popup-container"
>
  <div
    class="input-container"
  >
    <input
      class="input"
      id="notebook-name"
      placeholder="Введите имя для новой записной книжки"
      type="text"
    />
  </div>
  <label
    class="error-label"
    for="notebook-name"
  />
  <button
    class="primary-button notebook-create-btn"
  >
    Создать
    <div
      class="button-content"
    />
  </button>
</div>
`;
